src/main.rs

```diff
@@ -0,0 +1,135 @@
+#![feature(try_blocks)]
+
+mod config;
+mod loader;
+mod scraper;
+mod state;
+mod tg_bot;
+
+use self::config::Config;
+use self::loader::HtmlLoader;
+use self::scraper::{FromHTML, MainPage, Story};
+use self::state::State;
+use self::tg_bot::Bot;
+
+use handlebars::Handlebars;
+use serde_json::json;
+use std::time::Duration;
+
+const ALL_SIDES_MAINPAGE: &str = "https://www.allsides.com/unbiased-balanced-news";
+
+struct AllSidesTgImporter {
+    cfg: Config,
+    loader: HtmlLoader,
+    bot: Bot,
+    state: State,
+    template: Handlebars<'static>,
+}
+
+impl AllSidesTgImporter {
+    pub async fn try_new(cfg: Config) -> anyhow::Result<AllSidesTgImporter> {
+        let loader = HtmlLoader::try_new(&cfg.webdriver_host, cfg.webdriver_port).await?;
+        let bot = Bot::try_new(
+            &cfg.telegram_secret,
+            &cfg.telegram_channel,
+            &cfg.telegram_admin,
+        )?;
+        let state = State::try_new(&cfg.story_db)?;
+        let mut template = Handlebars::new();
+        template
+            .register_template_string("main", include_str!("../data/post-template.handlebars"))?;
+        Ok(AllSidesTgImporter {
+            cfg,
+            loader,
+            bot,
+            state,
+            template,
+        })
+    }
+
+    pub async fn run(mut self) -> anyhow::Result<()> {
+        let delay = Duration::from_secs(self.cfg.update_interval * 60);
+        loop {
+            if let Err(e) = self.tick().await {
+                log::error!("{}", e);
+                self.bot
+                    .log_error(e)
+                    .await
+                    .map_err(|e| log::error!("failed to post log message to telegram: {}", e))
+                    .ok();
+            }
+            tokio::time::delay_for(delay).await;
+        }
+    }
+
+    async fn tick(&mut self) -> anyhow::Result<()> {
+        let main_page = self.loader.open(ALL_SIDES_MAINPAGE).await?;
+        let main_page = MainPage::from_html(&main_page)?;
+        for teaser in main_page.teasers {
+            if self.state.is_published(&teaser.url)? {
+                continue;
+            }
+
+            let story = self.loader.open(&teaser.url).await?;
+            let story = Story::from_html(&story)?;
+            let formatted = self.format_story(&story, &teaser.url)?;
+
+            self.bot.publish_message(formatted).await?;
+            self.state.set_published(&teaser.url).await?;
+        }
+        Ok(())
+    }
+
+    fn format_story(&self, story: &Story, url: &str) -> anyhow::Result<String> {
+        let story_content = story
+            .summary
+            .iter()
+            .map(|p| p.telegram_html())
+            .collect::<Vec<String>>()
+            .join("\n\n"); // any way to do that nicely without collecting to Vec<String> first?
```

Comment:
`itertools::Itertools::join`. Although collecting into `Vec` has the benefit that `.join` can precalculate the needed memory exactly

Reply:
Well, in both cases, performance isn't of concern, but itertools option is cleaner. Thanks!